File: linux/amd/compiled/go/jump/jump_optimized.go
package main

import (
	"fmt"
	"math"
	"time"
)

func jumpSearchOptimized(arr []int, target int) int {
	n := len(arr)
	step := int(math.Sqrt(float64(n)))
	end := step

	for end < n && arr[end-1] < target {
		end += step
	}
	start := end - step
	if end > n {
		end = n
	}
	for i := start; i < end; i++ {
		if arr[i] == target {
			return i
		}
	}
	return -1
}

func main() {
	arr := []int{2, 4, 6, 8, 10, 12, 14, 16, 18}
	target := 14
	
	start := time.Now()
	index := jumpSearchOptimized(arr, target)
	duration := time.Since(start)

	fmt.Printf("Target found at index: %d\n", index)
	fmt.Printf("Execution time: %v\n", duration)
}

File: windows/intel/interpreted/ruby/jump/jump_optimized.rb
def jump_search_optimized(arr, target)
  n = arr.length
  step = Math.sqrt(n).to_i
  prev = 0
  max_index = [step, n].min - 1

  # Jump ahead in blocks
  while max_index < n && arr[max_index] < target
    prev = step
    step += Math.sqrt(n).to_i
    max_index = [step, n].min - 1
  end

  # Linear scan within block
  for i in prev..max_index
    return i if arr[i] == target
  end

  -1
end

# Test
arr = [2, 4, 6, 8, 10, 12, 14, 16, 18]
target = 14
index = jump_search_optimized(arr, target)
puts "Index: #{index}"
